### PR TITLE
Fixes for MSVC 2013

### PR DESCRIPTION
**Pull request by cboulay**
Here is a cleaner version of my changes to support MSVC 2013.
While everything builds fine, not of the programs really work without merging the changes from #149 and #162 If those are merged then all programs work, including utilities and examples.

I made a lot of changes to CMakeLists.txt
IMO, it's more organized now. It's probably (almost) possible to separate the tracker portions into a separate file, if that's desirable.

### CODE REVIEW

**Comment by cboulay** [comment]
Squashed and rebased to fix conflicts.

File: examples/c/test_record_video.c [hunk 1]
```diff
@@ -41,7 +41,8 @@
 
 int main(int arg, char** args) {
     int count = psmove_count_connected();
-    PSMove *moves[count];
+    //PSMove *moves[count];
```
Comment (thp) [inline examples/c/test_record_video.c]: Remove this line instead of commenting it out.

File: examples/c/test_roi_size.c [hunk 2]
```diff
@@ -104,16 +104,17 @@ save(int i, int roi_size, struct TestContext *context)
     cvSaveImage(path, frame, imgParams);
 }
 
+#define cROIS 3
```
Comment (thp) [inline examples/c/test_roi_size.c]: Named constant macros should be all-uppercase, maybe use `ROIS_COUNT`?

File: examples/c/test_roi_size.c [hunk 3]
```diff
@@ -104,16 +104,17 @@ save(int i, int roi_size, struct TestContext *context)
...
 {
     printf("\n -- PS Move API ROI Performance Test -- \n\n");
-    int roi_sizes[] = {480, 240, 120};
+    int roi_sizes[cROIS] = { 480, 240, 120 };
     int roi;
-    int rois = sizeof(roi_sizes)/sizeof(roi_sizes[0]);
-
-    float data[ITERATIONS][rois];
-    float position[ITERATIONS][rois][3]; // x, y, r
+    //int rois = sizeof(roi_sizes) / sizeof(roi_sizes[0]);  // MSVC is too stupid to figure out this is a constant. See cROIS above.
```
Comment (thp) [inline examples/c/test_roi_size.c]: Remove commented out code.

File: examples/c/test_roi_size.c [hunk 4]
```diff
@@ -124,12 +125,15 @@ main(int argc, char *argv[])
     putenv(PSMOVE_TRACKER_FILENAME_ENV "=test_roi_size.avi");
     putenv(PSMOVE_TRACKER_COLOR_ENV "=723a8c");
 
-    for (roi=0; roi<rois; roi++) {
+    for (roi = 0; roi<cROIS; roi++) {
         printf("Testing tracking performance: %d\n", roi_sizes[roi]);
 
-        char tmp[strlen(PSMOVE_TRACKER_ROI_SIZE_ENV) + 10];
+        //char tmp[strlen(PSMOVE_TRACKER_ROI_SIZE_ENV) + 10];  //msvc cannot handle variable length arrays, and is too stupid to figure out this is constant.
```
Comment (thp) [inline examples/c/test_roi_size.c]: Remove commented out code.

File: examples/c/test_tracker.c [hunk 5]
```diff
@@ -44,13 +44,15 @@
 int main(int arg, char** args) {
     int i;
     int count = psmove_count_connected();
-    PSMove* controllers[count];
+    //PSMove* controllers[count];
```
Comment (thp) [inline examples/c/test_tracker.c]: Remove commented out code.

File: external/README [hunk 6]
```diff
@@ -73,3 +73,24 @@ PS3EYEDriver
...
+https://www.sourceware.org/pthreads-win32/
+
+msvc-support
+============
+
+Work-arounds for MSVC's non-compliant C and a few unix-like libraries.
+These were taken from various sources.
+
+unistd.h is from http://stackoverflow.com/questions/341817/is-there-a-replacement-for-unistd-h-for-windows-visual-c
+There is no license.
```
Comment (thp) [inline external/README]: If there would be no license, we wouldn't be allowed to include it.

For stackoverflow, generally this applies: `user contributions licensed under cc by-sa 3.0 with attribution required`

Please spend some time figuring out from where the code came from and specify the license that applies or rewrite that part from scratch; we can't accept code without a license.

File: external/README [hunk 7]
```diff
@@ -73,3 +73,24 @@ PS3EYEDriver
...
+============
+
+A Windows implementation of posix-threads.
+https://www.sourceware.org/pthreads-win32/
+
+msvc-support
+============
+
+Work-arounds for MSVC's non-compliant C and a few unix-like libraries.
+These were taken from various sources.
```
Comment (thp) [inline external/README]: Please specify these "various sources".

Comment (thp) [inline external/README]: We also need license information here.

File: external/README [hunk 8]
```diff
@@ -73,3 +73,24 @@ PS3EYEDriver
...
+
+unistd.h is from http://stackoverflow.com/questions/341817/is-there-a-replacement-for-unistd-h-for-windows-visual-c
+There is no license.
+
+stdint.h is from http://snipplr.com/view.php?codeview&id=18199
+and is public domain.
+
+getopt.h is from https://gist.github.com/ashelly/7776712
+and uses the LGPL.
```
Comment (thp) [inline external/README]: "...uses the LGPL 2.1 (or later)".

File: src/psmove.c [hunk 9]
```diff
@@ -81,6 +85,33 @@
 #include "daemon/moved_client.h"
 #include "hidapi.h"
 
+/* Add gettimeofday for MSVC http://stackoverflow.com/a/26085827/1256069 */
+#ifdef _MSC_VER
+int gettimeofday(struct timeval * tp, struct timezone * tzp)
```
Comment (thp) [inline src/psmove.c]: Why not put this into msvc-support?

**Comment by thp** [comment]
Additional comments:

 * Please don't check in binaries (DLLs)
 * Put `ps3eye_settings.reg` somwhere below `contrib/`, not in the root directory

I'm still not sure whether we really need the whole pthreads-win32 library, just `#ifdef`'ing the relevant parts and calling the appropriate Win32 threading functions should do the trick; there's a slideshare that has a table with mappings:

http://www.slideshare.net/abufayez/pthreads-vs-win32-threads

**Comment by cboulay** [comment]
Thanks for the link about win32 threads. I'll look into it to see if it's an easy replacement and if so then I can get rid of pthreads-win32.

Some [changes I've made in one of my own branches](https://github.com/cboulay/psmoveapi/commit/e70330b9b0c7b064dc90a16fb8d61cd01d9cb47d) have eliminated the need for ps3eye_settings.reg entirely. The PS3EYE settings are saved to a file in the home directory then restored from there, just like in Linux.

I wonder though if this mechanism is even necessary (note OSX does not do it at all). From what I can tell, all that's being done is the original settings are being queried from the camera as soon as the camera is opened, the camera settings are manipulated by the API, then the original settings are restored to the camera before it is closed. In Windows, using the PS3EYEDriver, the 'original' settings are just the PS3EYEDriver defaults. There's no need to store these to a file. I propose we change these Win functions to no-ops, just like the Mac functions.

 Does it happen in Linux that if you change the camera settings with application A then those changed settings will be preserved when opening the camera with application B? If not, then these functions don't need to exist at all.

**Comment by thp** [comment]
Any news on this front?

**Comment by cboulay** [comment]
Sorry, I've had a backlog of other things. I can get back to this by the end of the week. Let me know if there's anything urgent and I can do a separate smaller pull request.

**Comment by cboulay** [comment]
I started over again in #182 .

**Comment by thp** [comment]
Closing in favor or #182 and consequently #190.

**Pull request closed without merging**